File: MatriX/samples/csharp/Weather/FrmMain.cs
using System;
using System.Drawing;
using System.Windows.Forms;

using Matrix;
using Matrix.Xml;
using Matrix.Xmpp;
using Matrix.Xmpp.Client;
using Weather.Settings;

namespace Weather
{
    public partial class FrmMain : Form
    {
        private Settings.Settings settings;
        private Login login;

        public FrmMain()
        {
            InitializeComponent();

            SetLicense();

            RegisterCustomElements();
            InitSettings();
        }

        private static void RegisterCustomElements()
        {
            Factory.RegisterElement<Weather>("ag-software:weather", "weather");

            Factory.RegisterElement<Settings.Settings>("ag-software:settings", "Settings");
            Factory.RegisterElement<Login>("ag-software:settings", "Login");
        }

        private void InitSettings()
        {
            settings = Util.LoadSettings();
            if (settings.Login == null)
                settings.Login = new Login();
            
            login = settings.Login;
            
            if (login != null)
            {
                if (!string.IsNullOrEmpty(login.User))
                    txtUsername.Text = login.User;

                if (!string.IsNullOrEmpty(login.Server))
                    txtServer.Text = login.Server;
                
                if (!string.IsNullOrEmpty(login.Password))
                    txtPassword.Text    = login.Password;
            }
        }

        /// <summary>
        /// Sets the license and activate the evaluation.
        /// </summary>
        private static void SetLicense()
        {
            /*
            const string LIC =@"your licence code we sent you";
            Matrix.License.LicenseManager.SetLicense(LIC);
            */
        }
       
        private void cmdConnect_Click(object sender, System.EventArgs e)
        {
            xmppClient.SetUsername(txtUsername.Text);
            xmppClient.SetXmppDomain(txtServer.Text);
            xmppClient.Password = txtPassword.Text;

            xmppClient.Status = "I'm chatty";
            xmppClient.Show = Matrix.Xmpp.Show.chat;

            // set settings
            login.User = txtUsername.Text;
            login.Server = txtServer.Text;
            login.Password = txtPassword.Text;

            
            xmppClient.Open();
        }

        private void xmppClient_OnError(object sender, ExceptionEventArgs e)
        {
            DisplayEvent("OnError");
        }

        private void xmppClient_OnLogin(object sender, Matrix.EventArgs e)
        {
            DisplayEvent("OnLogin");   
        }

        private void DisplayEvent(string ev)
        {
            listEvents.Items.Add(ev);
            listEvents.SelectedIndex = listEvents.Items.Count - 1;
        }

        private void xmppClient_OnBind(object sender, JidEventArgs e)
        {
            DisplayEvent("OnBind");
        }

        private void xmppClient_OnClose(object sender, Matrix.EventArgs e)
        {
            DisplayEvent("OnClose");
        }

        private void xmppClient_OnRosterStart(object sender, Matrix.EventArgs e)
        {

        }

        private void xmppClient_OnRosterEnd(object sender, Matrix.EventArgs e)
        {
            DisplayEvent("OnRosterEnd");
        }

        private void xmppClient_OnRosterItem(object sender, Matrix.Xmpp.Roster.RosterEventArgs e)
        {
            DisplayEvent(string.Format( "OnRosterItem\t{0}\t{1}", e.RosterItem.Jid, e.RosterItem.Name ));
        }

        private void xmppClient_OnPresence(object sender, PresenceEventArgs e)
        {
            DisplayEvent(string.Format("OnPresence\t{0}", e.Presence.From));
        }

        private void xmppClient_OnIq(object sender, IqEventArgs e)
        {
            DisplayEvent("OnIq");

            if (e.Iq.Type == IqType.get &&
                e.Iq.Query is Weather)
            {
                var weather = e.Iq.Query as Weather;
                string zip = weather.Zip;
                // here you should lookup the weather information for the given zip code in a database or webservice
                // we just return some random numbers
                
                var temp = new Random().Next(-10, 40);
                var humidity = new Random().Next(10, 90);

                var wiq = new WeatherIq
                              {
                                  To = e.Iq.From,
                                  Type = IqType.result,
                                  Weather = new Weather {Temperature = temp, Humidity = humidity}
                              };
                // send the response
                xmppClient.Send(wiq);
            }
        }

        private void xmppClient_OnMessage(object sender, MessageEventArgs e)
        {
            DisplayEvent("OnMessage");
        }

        private void xmppClient_OnValidateCertificate(object sender, CertificateEventArgs e)
        {
            // always accept cert
            e.AcceptCertificate = true;
        }
       
        private void cmdDisconnect_Click(object sender, System.EventArgs e)
        {
            xmppClient.Close();
        }

        private void xmppClient_OnAuthError(object sender, Matrix.Xmpp.Sasl.SaslEventArgs e)
        {
            DisplayEvent("OnAuthError");
            xmppClient.Close();
        }
       
        private void xmppClient_OnReceiveXml(object sender, TextEventArgs e)
        {
            rtfDebug.SelectionStart = rtfDebug.Text.Length;
            rtfDebug.SelectionLength = 0;
            rtfDebug.SelectionColor = Color.Red;
            rtfDebug.AppendText("RECV: ");
            rtfDebug.SelectionColor = Color.Black;
            rtfDebug.AppendText(e.Text);
            rtfDebug.AppendText("\r\n"); 
        }

        private void xmppClient_OnSendXml(object sender, TextEventArgs e)
        {
            rtfDebug.SelectionStart = rtfDebug.Text.Length;
            rtfDebug.SelectionLength = 0;
            rtfDebug.SelectionColor = Color.Blue;
            rtfDebug.AppendText("SEND: ");
            rtfDebug.SelectionColor = Color.Black;
            rtfDebug.AppendText(e.Text);
            rtfDebug.AppendText("\r\n");
        }
       
        private void xmppClient_OnStreamError(object sender, StreamErrorEventArgs e)
        {
            DisplayEvent("OnStreamError");
        }

      
        private void frmMain_FormClosed(object sender, FormClosedEventArgs e)
        {
            Util.SaveSettings(settings);
        }

        private void cmdRequestWeatherInfo_Click(object sender, System.EventArgs e)
        {
            if (!String.IsNullOrEmpty(txtZip.Text) && 
                !String.IsNullOrEmpty(txtRequestFrom.Text))
                RequestWeatherInfo(txtRequestFrom.Text, txtZip.Text);
        }

        private void RequestWeatherInfo(Jid from, string zip)
        {
            var wiq = new WeatherIq
                          {
                              Type = IqType.get,
                              To = from,
                              Weather = new Weather { Zip = zip }
                          };
            // we pass the zip code as state object to the IqFilter
            xmppClient.IqFilter.SendIq(wiq, WeatherInfoResponse, zip);
        }

        private void WeatherInfoResponse(object sender, IqEventArgs e)
        {
            var iq = e.Iq;

            if (iq.Type == IqType.result)
            {
                var weather = iq.Element<Weather>();
                if (weather != null)
                {
                    var zip = e.State as string;

                    MessageBox.Show(String.Format("weather info for zip code: {0}\r\n\r\nHumidity: {1}\r\nTemperature: {2}",
                                                  zip, weather.Humidity, weather.Temperature));
                }
            }
            else if (iq.Type == IqType.error)
            {
                // process errors here
            }
        }
    }
}

File: MatriX/samples/vb/Weather/FrmMain.vb
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Imports Matrix
Imports Matrix.Xml
Imports Matrix.Xmpp
Imports Matrix.Xmpp.Client
Imports WeatherVB.Settings

Public Partial Class FrmMain
	Inherits Form
	Private settings As Settings.Settings
	Private login As Login

	Public Sub New()
        InitializeComponent
		SetLicense()

		RegisterCustomElements()
		InitSettings()
	End Sub

	Private Shared Sub RegisterCustomElements()
		Factory.RegisterElement(Of Weather)("ag-software:weather", "weather")

		Factory.RegisterElement(Of Settings.Settings)("ag-software:settings", "Settings")
		Factory.RegisterElement(Of Login)("ag-software:settings", "Login")
	End Sub

	Private Sub InitSettings()
		settings = Util.LoadSettings()
		If settings.Login Is Nothing Then
			settings.Login = New Login()
		End If

		login = settings.Login

		If login IsNot Nothing Then
			If Not String.IsNullOrEmpty(login.User) Then
				txtUsername.Text = login.User
			End If

			If Not String.IsNullOrEmpty(login.Server) Then
				txtServer.Text = login.Server
			End If

			If Not String.IsNullOrEmpty(login.Password) Then
				txtPassword.Text = login.Password
			End If
		End If
	End Sub

	''' <summary>
	''' Sets the license and activate the evaluation.
	''' </summary>
	Private Shared Sub SetLicense()
		'
        ' const string LIC =@"your licence code we sent you";
        ' Matrix.License.LicenseManager.SetLicense(LIC);
        '    
    End Sub

	Private Sub cmdConnect_Click(sender As Object, e As System.EventArgs) Handles cmdConnect.Click
		xmppClient.SetUsername(txtUsername.Text)
		xmppClient.SetXmppDomain(txtServer.Text)
		xmppClient.Password = txtPassword.Text

		xmppClient.Status = "I'm chatty"
		xmppClient.Show = Xmpp.Show.chat

		' set settings
		login.User = txtUsername.Text
		login.Server = txtServer.Text
		login.Password = txtPassword.Text


		xmppClient.Open()
	End Sub

	Private Sub xmppClient_OnError(sender As Object, e As ExceptionEventArgs) Handles xmppClient.OnError
		DisplayEvent("OnError")
	End Sub

	Private Sub xmppClient_OnLogin(sender As Object, e As Matrix.EventArgs) Handles xmppClient.OnLogin
		DisplayEvent("OnLogin")
	End Sub

	Private Sub DisplayEvent(ev As String)
		listEvents.Items.Add(ev)
		listEvents.SelectedIndex = listEvents.Items.Count - 1
	End Sub

	Private Sub xmppClient_OnBind(sender As Object, e As JidEventArgs) Handles xmppClient.OnBind
		DisplayEvent("OnBind")
	End Sub

	Private Sub xmppClient_OnClose(sender As Object, e As Matrix.EventArgs) Handles xmppClient.OnClose
		DisplayEvent("OnClose")
	End Sub

	Private Sub xmppClient_OnRosterStart(sender As Object, e As Matrix.EventArgs) Handles xmppClient.OnRosterStart
        DisplayEvent("OnRosterStart")
	End Sub

	Private Sub xmppClient_OnRosterEnd(sender As Object, e As Matrix.EventArgs) Handles xmppClient.OnRosterEnd
		DisplayEvent("OnRosterEnd")
	End Sub

	Private Sub xmppClient_OnRosterItem(sender As Object, e As Matrix.Xmpp.Roster.RosterEventArgs) Handles xmppClient.OnRosterItem
		DisplayEvent(String.Format("OnRosterItem" & vbTab & "{0}" & vbTab & "{1}", e.RosterItem.Jid, e.RosterItem.Name))
	End Sub

	Private Sub xmppClient_OnPresence(sender As Object, e As PresenceEventArgs) Handles xmppClient.OnPresence
		DisplayEvent(String.Format("OnPresence" & vbTab & "{0}", e.Presence.From))
	End Sub

	Private Sub xmppClient_OnIq(sender As Object, e As IqEventArgs) Handles xmppClient.OnIq
		DisplayEvent("OnIq")

		If e.Iq.Type = IqType.[get] AndAlso TypeOf e.Iq.Query Is Weather Then
			Dim weather = TryCast(e.Iq.Query, Weather)
			Dim zip As String = weather.Zip
			' here you should lookup the weather information for the given zip code in a database or webservice
			' we just return some random numbers

			Dim temp = New Random().[Next](-10, 40)
			Dim humidity = New Random().[Next](10, 90)

			Dim wiq = New WeatherIq() With { _
				.[To] = e.Iq.From, _
				.Type = IqType.result, _
				.Weather = New Weather() With { _
					.Temperature = temp, _
					.Humidity = humidity _
				} _
			}
			' send the response
			xmppClient.Send(wiq)
		End If
	End Sub

	Private Sub xmppClient_OnMessage(sender As Object, e As MessageEventArgs) Handles xmppClient.OnMessage
		DisplayEvent("OnMessage")
	End Sub

	Private Sub xmppClient_OnValidateCertificate(sender As Object, e As CertificateEventArgs) Handles xmppClient.OnValidateCertificate
		' always accept cert
		e.AcceptCertificate = True
	End Sub

	Private Sub cmdDisconnect_Click(sender As Object, e As System.EventArgs) Handles cmdDisconnect.Click
		xmppClient.Close()
	End Sub

	Private Sub xmppClient_OnAuthError(sender As Object, e As Sasl.SaslEventArgs) Handles xmppClient.OnAuthError
		DisplayEvent("OnAuthError")
		xmppClient.Close()
	End Sub

	Private Sub xmppClient_OnReceiveXml(sender As Object, e As TextEventArgs) Handles xmppClient.OnReceiveXml
		rtfDebug.SelectionStart = rtfDebug.Text.Length
		rtfDebug.SelectionLength = 0
		rtfDebug.SelectionColor = Color.Red
		rtfDebug.AppendText("RECV: ")
		rtfDebug.SelectionColor = Color.Black
		rtfDebug.AppendText(e.Text)
		rtfDebug.AppendText(vbCr & vbLf)
	End Sub

	Private Sub xmppClient_OnSendXml(sender As Object, e As TextEventArgs) Handles xmppClient.OnSendXml
		rtfDebug.SelectionStart = rtfDebug.Text.Length
		rtfDebug.SelectionLength = 0
		rtfDebug.SelectionColor = Color.Blue
		rtfDebug.AppendText("SEND: ")
		rtfDebug.SelectionColor = Color.Black
		rtfDebug.AppendText(e.Text)
		rtfDebug.AppendText(vbCr & vbLf)
	End Sub

	Private Sub xmppClient_OnStreamError(sender As Object, e As StreamErrorEventArgs) Handles xmppClient.OnStreamError
		DisplayEvent("OnStreamError")
	End Sub
    
	Private Sub cmdRequestWeatherInfo_Click(sender As Object, e As System.EventArgs) Handles cmdRequestWeatherInfo.Click
		If Not String.IsNullOrEmpty(txtZip.Text) AndAlso Not [String].IsNullOrEmpty(txtRequestFrom.Text) Then
			RequestWeatherInfo(txtRequestFrom.Text, txtZip.Text)
		End If
	End Sub

	Private Sub RequestWeatherInfo(from As Jid, zip As String)
		Dim wiq = New WeatherIq() With { _
			.Type = IqType.[get], _
			.[To] = from, _
			.Weather = New Weather() With { _
				.Zip = zip _
			} _
		}
		' we pass the zip code as state object to the IqFilter
		xmppClient.IqFilter.SendIq(wiq,AddressOf WeatherInfoResponse, zip)
	End Sub

	Private Sub WeatherInfoResponse(sender As Object, e As IqEventArgs)
		Dim iq = e.Iq

		If iq.Type = IqType.result Then
			Dim weather = iq.Element(Of Weather)()
			If weather IsNot Nothing Then
				Dim zip = TryCast(e.State, String)

				MessageBox.Show([String].Format("weather info for zip code: {0}" & vbCr & vbLf & vbCr & vbLf & "Humidity: {1}" & vbCr & vbLf & "Temperature: {2}", zip, weather.Humidity, weather.Temperature))
			End If
				' process errors here
		ElseIf iq.Type = IqType.[error] Then
		End If
	End Sub

  
    Private Sub FrmMain_FormClosing( sender As System.Object,  e As System.Windows.Forms.FormClosingEventArgs) Handles MyBase.FormClosing
            Util.SaveSettings(settings)
    End Sub
end Class

